Core functionality for version 1.2 (#58)
beacon-biosignals/StableHashTraits.jl: README.md

StableHashTraits.transformer(::Type{<:MyType}) = Transformer(x -> (; x.data);
                                                             preserves_structure=true)
a = MyType(read("myfile.txt"), Dict{Symbol, Any}(:read => Dates.now()))
b = MyType(read("myfile.txt"), Dict{Symbol, Any}(:read => Dates.now()))
stable_hash(a; version=3) == stable_hash(b; version=3) # true
```
<!--END_EXAMPLE-->

Useres can define a method of `transformer` to customize how an object is hashed. It should return a function wrapped in `Transformer`. During hashing, this function is called and its result is the value that is actually hashed. (The `preserves_structure` keyword shown above is an optional flag that can be used to further optimize performance of your transformer in some cases; you can do this any time the function is type stable, but some type-unstable functions are also possible. See the documentation for details).

StableHashTraits aims to guarantee a stable hash so long as you only upgrade to non-breaking versions (e.g. `StableHashTraits = "1"` in `[compat]` of `Project.toml`); any changes in an object's hash in this case would be considered a bug.

> [!WARNING]

> > Hash versions 3 constitutes a substantial redesign of StableHashTraits so as to avoid reliance on some unstable Julia internals. Hash versions 1 and 2 are deprecated and will be removed in a soon-to-be released StableHashTraits 2.0. Hash version 3 will remain unchanged in this 2.0 release. Hash version 1 is the default version if you don't specify a version.

Reviewer: ```suggestion
> Hash versions 3 constitutes a substantial redesign of StableHashTraits so as to avoid reliance on some unstable Julia internals. Hash versions 1 and 2 are deprecated and will be removed in a soon-to-be released StableHashTraits 2.0. Hash version 3 will remain unchanged in this 2.0 release. Hash version 1 is the default version if you don't specify a version. To see the documentation for hash version 1, go here-TODO
```

<!--The START_ and STOP_ comments are used to extract content that is also repeated in the documentation-->
<!--START_OVERVIEW-->
## Use Case and Design Rationale

StableHashTraits is designed to be used in cases where there is an object you wish to serialize in a content-addressed cache. How and when objects pass the same input to a hashing algorithm is meant to be predictable and well defined, so that the user can reliably define methods of `transformer` to modify this behavior.

## What gets hashed?

By default, an object is hashed according to its `StructType` (ala
[SructTypes](https://github.com/JuliaData/StructTypes.jl)), and this can be customized using
[`StableHashTraits.transformer`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transformer).

> Hashing makes use of [`stable_type_name`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.stable_type_name) which generates `string(parentmodule(T)) * "." * string(nameof(T))` for type `T`, with a few additional regularizations to ensure e.g. `Core.` values become `Base.` values (as what is in `Core` vs. `Base` changes across julia versions). This function also ensures that no anonymous values (those that include `#`) are hashed, as these are not stable across sessions.

Reviewer: Proposal: very conservative by default, but easy to opt-in to different behavior by defining `StableHashTraits.transformer` methods:

* types by default error in value position
* functions by default error 
* singleton types by default error

Examples:

```julia
f(x) = x+1
# ideally: per-function
StableHashTraits.transformer(::typeof(f), ::MyContext) = Transformer(ParentModuleNameOfFieldHasher)

# if you want to, can opt-in for all functions, but can be risky (if they move submodules then hash changes):
StableHashTraits.transformer(f::T, ::MyContext) where {T <: Function} = Transformer(ParentModuleNameOfFieldHasher)

# can support types or a family of types or all types:
StableHashTraits.transformer(::Type{T}, ::MyContext) where {T <: MyAbstractType} = Transformer(ParentModuleNameOfHasher)
```

Rationale:
* hash implementers unfortunately need to think carefully about how to hash tricky things like types and functions, and when the hash should or should not change, so error by default
* packages that want to support StableHashTraits first-class can define `StableHashTraits.transformer` methods (without a context) for their functions/types, to declare how they should be hashed (and they should test their hashes don't change when they don't want them to)
* users can define `StableHashTraits.transformer` with a context to define how upstream functions/types are hashed (without piracy)

- `Type`: when hashing the type of an object or its contained types, only the name of `stable_type_name(StructType(T))` is hashed along with any structure as determined by the particular return value of `StructType(T)` (e.g. `eltype` for `ArrayType`). If you hash a type as a value (e.g. `stable_hash(Int)`) the `stable_type_name` of the type itself, rather than `StructType(T)` is used.

> - `StructType.DataType` — the `fieldnames`, `fieldtypes` and the field values are hashed, and if this is a `StructType.UnorderedStruct` those are all sorted in lexicographic order of the fieldnames. `StructType.UnorderedStruct` is the default struct-type trait so this is how most objects get hashed.

Reviewer: let's consider adding `nameof(::Type{MyStruct})` into the hash here, without `parentmodule`, just the name of the struct type as a symbol

- `StructType.ArrayType` — the `eltype` is hashed and elements are hashed using `iterate`.
If [`StableHashTraits.is_ordered`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.is_ordered) returns `false` the elements are first `sort`ed according to [`StableHashTraits.hash_sort_by`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.hash_sort_by).

- `StructType.DictType` — the `eltype` of the keys and values are hashed by iterating over `StructTypes.keyvaluepairs`. If [`StableHashTraits.is_ordered`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.is_ordered) returns `false` the pairs of the dictionary are first `sort`ed according their keys using [`StableHashTraits.hash_sort_by`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.hash_sort_by).

- `StructType.CustomStruct` - the object is first `StructType.lower`ed and the result is hashed according to its `StructType`.

> - `StructType.NullType`, `StructType.SingletonType`: in these cases the `stable_type_name` of the type `T` is hashed rather than `StructType(T)`

Reviewer: `nothing` and `missing` we explicitly define a method (they hash differently). For other nulltypes/singleton types the user must define how they should hash (error by default).

> - `StructType.NumberType`, `StructType.StringType`, `StructType.BoolType`: the the type of the object is hashed along with its bytes

Reviewer: we should be explicit that it's the `StructType.NumberType` that gets hashed, not e.g. `Float64`

- `Function`: functions are a special case and their `stable_type_name` is hashed along with their fieldnames, fieldtypes and fieldvalues. Functions have fields when they are curried, e.g. `==(2)` or when they are defined via a `struct` definition.

> - `AbstractRange`: though their `StructType` is `AbstractArray`, abstract ranges are treated as `UnorderedStruct` objects for purposes of hashing, as this normally leads to a more effeicient hash computation.

Reviewer: We will hash these as an `OrderedStruct` with fields `(first, step, last)` with the token `StableHashTraits.RangeType` to disambiguate from structs like:
```julia
struct MyObj
   first
   step
   last
end
```

So basically we are treating `AbstractRange` as another first-class object like arrays/structs/numbers/etc

## What gets hashed? (hash version 4)

This covers the behavior when using the latest hash version (4). You can read the documentation for hash version 1 [here](https://github.com/beacon-biosignals/StableHashTraits.jl/blob/v1.0.0/README.md) and hash version 2-3 [here](https://github.com/beacon-biosignals/StableHashTraits.jl/blob/v1.1.8/README.md).

When you call `stable_hash(x; version=4)`, StableHashTraits hashes both the value `x` and its type `T`, normally doing so according to `StructType(T)`. (ala
[SructTypes](https://github.com/JuliaData/StructTypes.jl)).

You can customize how the value is hashed using [`StableHashTraits.transformer`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transformer),
and how its type is hashed using [`StableHashTraits.transform_type`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transform_type).
If you need to customize either of these functions for a type that you don't own, you can use a [@context](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.@context) to avoid type piracy.

> ### `StructType.DataType`

Reviewer: ```suggestion
### `StructType.DataType`

This represents a "record" type whose fields include data. This is the default struct type.

```

> To hash the value, each field value of a data type is hashed.

Reviewer: from pair review- discussed using code here (`fieldnames`) for clarity

The type of a data type is hashed using `string(nameof(T))`, the `fieldnames(T)`, (sorting them for `UnorderedStruct`), along with a hash of the type of each element of `fieldtypes(T)` according to their `StructType`.

No type parameters are hashed by default. To hash these you need to specialize on [`transform_type`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transform_type) for your struct. Note that because `fieldtypes(T)` is hashed, you don't need to do this unless your type parameters are not used in the specification of your field types.

### `StructType.ArrayType`

To hash the value, each element of an array type is hashed using `iterate`. If the object `isa AbstractArray`, the `size(x)` of the object is also hashed.

If [`StableHashTraits.is_ordered`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.is_ordered) returns `false` the elements are first `sort`ed according to [`StableHashTraits.hash_sort_by`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.hash_sort_by).

> To hash the type, the string `"StructTypes.ArrayTYpe"` is hashed (meaning that the kind of array won't matter to the hash value), and the type of the `elype` is hashed, according to its `StructType`. If the type `<: AbstractArray`, the `ndims(T)` is hashed.

Reviewer: ```suggestion
To hash the type, the string `"StructTypes.ArrayType"` is hashed (meaning that the kind of array won't matter to the hash value), and the type of the `elype` is hashed, according to its `StructType`. If the type `<: AbstractArray`, the `ndims(T)` is hashed.
```

If [`StableHashTraits.is_ordered`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.is_ordered) returns `false` (which is the default return value) the pairs are first `sort`ed according their keys using [`StableHashTraits.hash_sort_by`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.hash_sort_by).

To hash the type, the string `"StructTypes.DictType"` is hashed (meaning that the kind of dictionary won't matter), and the type of the `keytype` and `valtype` is hashed, according to its `StructType`.

### `AbstractRange`

`AbstractRange` constitutes an exception to the rule that we use `StructType`: for efficient hashing, ranges are treated as another first-class container object, separate from array types.

The value is hashed as `(first(x), step(x), last(x))`.

The type is hashed as `"Base.AbstractRange"` along with the type of the `eltype`, according to its `StructType`. Thus, the type of range doesn't matter (just that it is a range).

### `StructTypes{Number/String/Bool}Type`

To hash the value, the result of `Base.write`ing the object is hashed.

> To hash the type, the value of `module_nameof_string(StructType))` is used. [`module_nameof_string`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.parentmodule_name) can be used to provide a name for an object composed of `string(parentmodule(T))` and `string(nameof(T))`, along with some additional regularizations and validations. Note that many packages consider the module of a type to be an implementation detail, and that this can change in a non-breaking release of a package.

Reviewer: ```suggestion
To hash the type, the value of `string("StructType.", nameof_string(StructType(T))))` is used. Note this means the type of the value itself is not being hashed, rather a string related to its struct type.
```

### `StructType.CustomStruct`

For any `StructType.CustomStruct`, the object is first `StructType.lower`ed and the result is hashed according to the lowered `StructType`.

### `missing` and `nothing`

> There is no value hashed for `missing` or `nothing`; the type is hashed as the string `"Base.Missing"` and `"Base.Nothing"` respectively.

Reviewer: ```suggestion
There is no value hashed for `missing` or `nothing`; the type is hashed as the string `"Base.Missing"` and `"Base.Nothing"` respectively. Note in particular the string `"Base.Missing"` does not have the same hash as `missing`, since the former would have it's struct type hashed.
```

### `StructType.{Null/Singleton}Type`

> These types error by default. You can opt in to behavior similar to `missing` and `nothing` by defining an appropriate [`transform_type`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.type_identifier).

Author: Alternative: use `nameof_string` for the type. This should be quite stable (since we're not relying on the module)

If `transform_type` for a singleton is defined, it will no longer error as the value is already defined to be empty.

### `Function`

> Attempting to hash a function errors by default. You can opt in to hashing a function

Author: Alternative: use `nameof_string` since it should be quite stable.

> These types error by default. You can opt in to behavior similar to `missing` and `nothing` by defining an appropriate [`transform_type`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.type_identifier) or by using [HashNullTypes](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.HashNullTypes) and/or [HashSingletonTypes](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.HashSingletonTypes).

Reviewer: from pair review: we plan to use `nameof` to provide a default (`nameof` for the type with nothing for the value)

by defining [`transform_type`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.type_identifier). If `transform_type` is defined, functions will hash as if they were `UnorderedStructs`: functions can have fields if they are curried (e.g. `==(2)`), and so, for this reason, the fields are included in the hash by default.

If a function value `fn` can be hashed in this way, so can types of the form `typeof(fn)`.

### `Type`

> When a type is provided as a value (e.g. `stable_hash(Int; version=4)`) hashing it will error by default.

Author: ditto?

You can opt in to hashing a given type as a value by defining [`transform_type_value`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transform_type_value) for your type, or by using [HashTypeValues](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.HashTypeValues)

## Examples

All of the following hash examples follow directly from the definitions above, but may not be so obvious to the reader.

> Most of the behaviors described below can be customized/changed by using your own hash [`@context`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.@context), which can be passed as the second argument to [`stable_hash`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.stable_hash).

Reviewer: ```suggestion
Most of the behaviors described below can be customized/changed by using your own hash [`@context`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.@context), which can be passed as the second argument to [`stable_hash`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.stable_hash). StableHashTraits tries to defer to StructTypes for most defaults instead of making more opinionated choices.
```

> The order of NamedTuple pairs does not matter

Reviewer: ```suggestion
The order of NamedTuple pairs does not matter, because StructTypes considers them as `UnorderedStruct`
```

```julia
stable_hash((;a=1,b=2); version=4) == stable_hash((;b=2,a=1); version=4)
```

> Two structs with the same fields and name hash equivalently

Reviewer: ```suggestion
Two structs with the same fields and name hash equivalently, because StructTypes considers them as `UnorderedStruct`
```

```julia
module A
    struct X
        bar::Int
        foo::Float64
    end
end

module B
    struct X
        foo::Float64
        bar::Int
    end
end

stable_hash(B.X(2, 1); version=4) == stable_hash(A.X(1, 2); version=4)
```

Different array types with the same content hash to the same value.

```julia
stable_hash(view([1,2,3], 1:2); version=4) == stable_hash([1,2]; version=4)
```

Byte equivalent arrays of all `NumberType` values will hash to the same value.

```julia

> stable_hash([0.0, 0.0]; version=4) == stable_hash([0, 0]; version=4)

Reviewer: We should add this example:
```julia
stable_hash([0.0f0, 0.0f0]; version=4) != stable_hash([0, 0]; version=4)
```

Also, even though the bytes are the same, since the size is hashed, we have:
```julia
stable_hash([0.0f0, 0.0f0]; version=4) != stable_hash([0]; version=4)
```
